File: BasicCalculatorII2.py
class Solution:
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        i = 0
        cur = 0
        result = 0
        sign = '+'
        while i < len(s):
            if s[i].isdigit():
                j = i + 1
                while j < len(s) and s[j].isdigit():
                    j += 1
                num = int(s[i : j])
                i = j - 1
                if sign == '+':
                    cur += num
                elif sign == '-':
                    cur -= num
                elif sign == '*':
                    cur *= num
                elif sign == '/':
                    if cur > 0:
                        cur = cur // num
                    else:
                        cur = -((-cur) // num)
            elif s[i] != ' ':
                if s[i] == '+' or s[i] == '-':
                    result += cur
                    cur = 0
                sign = s[i]
            i += 1
        return result + cur
        
if __name__ == '__main__':
    s = raw_input()
    solution = Solution()
    print(solution.calculate(s))

File: BasicCalculatorII2.cpp
#include<iostream>
#include<string>
#include<vector>

using namespace std;

class Solution {
public:
    int calculate(string s) {
        
        vector<int> nums;
        
        int num = 0;
        char sign = '+';
        
        for(int i=0;i<s.length();i++){
            char c = s[i];
            //if(c == ' ') continue;
            if(c >= '0' && c <= '9'){
                num = num * 10 + s[i] - '0';
            }
            if(c == '+' || c == '-' || c == '*' || c == '/' || i == s.length() - 1){
                if(sign == '+') nums.push_back(num);
                if(sign == '-') nums.push_back(-num);
                if(sign == '*'){
                    num = nums.back() * num;
                    nums.pop_back();
                    nums.push_back(num);
                }
                if(sign == '/') {
                    num = nums.back() / num;
                    nums.pop_back();
                    nums.push_back(num);
                }
                sign = c;
                num = 0;
            }
        }
        
        int result = 0;
        for(int i=0;i<nums.size();i++){
            result += nums[i];
        }
        
        return result;
    }
};

int main(){
    
    string s;
    getline(cin, s);
    
    Solution *solution = new Solution();
    cout<<solution->calculate(s);
    
    return 0;
    
}
